cosmoz-image-viewer.js

```diff
@@ -344,52 +352,123 @@
 			}
 
 			if (this.hasWindow) {
-				globals.window.document.title = this._detachedWindowTitle;
-				globals.window.ciw.setImages(this._resolvedImages, this.currentImageIndex);
-				globals.window.focus();
-				return;
+				return this._detachToExistingWindow();
 			}
 
+			globals.windowBeforeUnloadHandler = () => {
+				if (globals.windowOpener !== null) {
+					globals.windowOpener._setIsDetached(false);
+				}
+				globals.windowOpener = null;
+				globals.window = null;
+			};
+
+			return this._detachToNewWindow();
+		},
+
+		_detachToNewWindow() {
 			const
 				w = globals.window = window.open(undefined, this.detachedWindowName, this._detachedWindowFeaturesString),
 				windowTemplate = this.$$('#externalWindow'),
-				windowTemplateClone = windowTemplate.content.cloneNode(true);
+				windowTemplateClone = windowTemplate.content.cloneNode(true),
+				setImages = () => w.ciw.setImages(this._resolvedImages, this.currentImageIndex);
 
 			if (w == null) {
 				// if window.open() is blocked (popup blocked, not emited by native user triggered event)
-				return;
+				return w;
 			}
 
 			if (w.ciw != null) {
 				// browser has been refreshed without closing external window, we will reuse an old window
 				w.removeEventListener('beforeunload', globals.windowBeforeUnloadHandler);
 			}
 
-			globals.windowBeforeUnloadHandler = () => {
-				globals.windowOpener._setIsDetached(false);
-				globals.windowOpener = null;
-				globals.window = null;
-			};
-
 			w.document.title = this._detachedWindowTitle;
 
+			w.addEventListener('download', ({detail}) => this.createZipFromUrls(detail).then(zip => this.downloadZip(zip)));
 			w.addEventListener('beforeunload', globals.windowBeforeUnloadHandler);
 
 			if (w.ciw == null) {
-				w.addEventListener('ready', () => w.ciw.setImages(this._resolvedImages, this.currentImageIndex));
-				w.document.body.appendChild(windowTemplateClone);
+				w.addEventListener('ready', () => setImages());
+
+				Array.from(windowTemplateClone.childNodes)
+					.forEach(node => {
+						if (node.tagName === 'SCRIPT') {
+							// Needed for Firefox
+							// otherwise the script would not be evaluated
+							const sc = document.createElement('script');
+							sc.innerHTML = node.innerHTML;
+							w.document.body.appendChild(sc);
+							return;
+						}
+						w.document.body.appendChild(node);
+					});
 			} else {
-				w.ciw.setImages(this._resolvedImages, this.currentImageIndex);
+				setImages();
 			}
+
+			return w;
 		},
+
+		_detachToExistingWindow() {
+			globals.window.document.title = this._detachedWindowTitle;
+			globals.window.ciw.setImages(this._resolvedImages, this.currentImageIndex);
+			globals.window.focus();
+			return globals.window;
+		},
+
 		get hasWindow() {
-			return globals.window != null && !globals.window.closed;
+			return globals.window != null && !globals.window.closed && globals.window.ciw;
```

Comment:
Why do we need this extra check?

Reply:
Thx, was needed before fixing firefox script evaluation

Reply:
Hmm ok, it is bad changing code to make tests pass, but the implications should be limited in this case I guess.
Could we theoretically end up in a scenario where we will unexpectedly reuse a window that hasn't got `ciw` and cause errors in other places? I guess another component or part of the app could open a window with the same title and we get that one when we call `window.open` ?

Reply:
True, that is the only case I can think of. Unlikely but possible. That would justify the check :)